docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,319 @@
+
+#### <a name='Story3-DeletinganindividualAzureMachinePoolMachine'></a>Story 3 - Deleting an individual Azure Machine Pool Machine
+
+Kai is an engineer in a large organization which has a MachinePool running. Kai needs to delete the Machine Pool. Kai uses `kubectl` to delete the Machine Pool. After a few seconds, Kai checks the Conditions on the MachinePool and sees that the VM is being deleted.
+
+### <a name='ImplementationDetailsNotesConstraints'></a>Implementation Details/Notes/Constraints
+
+There is an existing implementation of asynchronous reconciliation for AzureMachinePools. The   `AzureMachinePoolStatus` stores a single `LongRunningOperationState` used to keep the Future returned by VMSS long running operations.
+
+```go
+// Future contains the data needed for an Azure long-running operation to continue across reconcile loops.
+type Future struct {
+    // Type describes the type of future, update, create, delete, etc
+    Type string `json:"type"`
+    // ResourceGroup is the Azure resource group for the resource
+    // +optional
+    ResourceGroup string `json:"resourceGroup,omitempty"`
+    // Name is the name of the Azure resource
+    // +optional
+    Name string `json:"name,omitempty"`
+    // FutureData is the base64 url encoded json Azure AutoRest Future
+    FutureData string `json:"futureData,omitempty"`
+}
+
+// AzureMachinePoolStatus defines the observed state of AzureMachinePool
+AzureMachinePoolStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+    
+    // LongRunningOperationState saves the state for an Azure long-running operations so it can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationState *infrav1.Future `json:"longRunningOperationState,omitempty"`
+}
+```
+
+### <a name='ProposedAPIChanges'></a>Proposed API Changes
+
+The proposed changes below show the changes to AzureCluster, AzureMachine, AzureMachinePool, and AzureMachinePoolMachine. The existing `LongRunningOperationState` field in AzureMachinePoolStatus will be pluralized to `LongRunningOperationStates` to store a list of Futures, following a similar pattern than Conditions, and will be extended to other CAPZ CRDs. In addition, the `Name` field of the `Future` type will be made required, as it becomes the identifier for a Future.
+
+```go
+// Future contains the data needed for an Azure long-running operation to continue across reconcile loops.
+type Future struct {
+    // Type describes the type of future, update, create, delete, etc
+    Type string `json:"type"`
+    // ResourceGroup is the Azure resource group for the resource
+    // +optional
+    ResourceGroup string `json:"resourceGroup,omitempty"`
+    // Name is the name of the Azure resource
+    Name string `json:"name"`
```

Comment:
Is this going to be used to identify the state of a long running service? As in, if there is a service with 3 resources, there will be 3 future objects with the name being the unique identifier. Is my understanding correct? 
If that's the case, should we also add a service name so that we can filter by service? It might be useful when we are updating conditions.

Reply:
>  As in, if there is a service with 3 resources, there will be 3 future objects with the name being the unique identifier. Is my understanding correct?

Correct

> If that's the case, should we also add a service name so that we can filter by service? It might be useful when we are updating conditions.

that's a good idea, will add.